### ISSUE TITLE
TTS Latency Occurs During Normal Navigation But Disappears When an NVDA Dialog is Open

### ISSUE THREAD

**Issue by lbk2907** [nvaccess/nvda#19551]
### A Brief Description
There is noticeable TTS (Text-to-Speech) latency/lag during general navigation. This lag is present across various synthesizers but is most prominent when using SAPI5 voices with slight inherent delay, such as Blastbay Libby or Blastbay Richard.

### Steps to reproduce
1. Use any TTS and voice (ideally Blastbay Libby or Blastbay Richard from Blastbay Studios, available to download at https://blastbay.com/voices.php).
2. Navigate through the OS (e.g., Desktop icons or the file list in File Explorer quickly) and observe the slight delay between keypress and speech output.
3. Open an NVDA dialog (e.g., press NVDA+Ctrl+S for the Select Synthesizer dialog).
4. Alt+Tab away from that dialog to another application (like File Explorer).
5. Navigate again to the file list quickly and observe that the speech response is now instantaneous and the lag is gone.

### Actual behavior
TTS response has a small but noticeable latency during normal use. This latency vanishes only when an NVDA GUI element is active or open in the background.

### Expected behavior
TTS response should be consistently fast and exhibit the same low latency regardless of whether an NVDA dialog is open or closed.

### NVDA logs, crash dumps and other attachments
No specific error in logs; this is a performance/latency observation. Note: The voices used for testing are commercial SAPI5 voices from Blastbay Studios.

#### NVDA installed/portable/running from source
Install and Portable

#### NVDA version
2025.3.2 (2025.3.2.53588)

#### Windows version
Windows 11 25H2 (AMD64) build 26200.7705

#### Name and version of other software in use when reproducing the issue
Blastbay Libby and Blastbay Richard (SAPI5 voices)

#### Other information about your system
Processor: Intel(R) Core(TM) 5 210H, 2200 Mhz, 8 Core(s), 12 Logical Processor(s)
Installed Physical Memory (RAM): 32.0 GB
RAM Slot 1: 16.0 GB, 5200 MT/s
RAM Slot 2: 16.0 GB, 5200 MT/s
Graphiccard: NVIDIA GeForce RTX 4050

#### Does the issue still occur after restarting your computer?
Yes

#### Have you tried any other versions of NVDA? If so, please report their behaviors
No

#### If NVDA add-ons are disabled, is your problem still occurring?
Yes

#### Does the issue still occur after you run the COM Registration Fixing Tool in NVDA's tools menu?
Yes

**Comment by SaschaCowley** [comment]
Welcome @lbk2907,

We are unable to process this issue as it stands.

- Please reproduce this issue and provide a [log file](https://github.com/nvaccess/nvda/wiki/LogFilesAndCrashDumps) of the behaviour. Ensure your log level is set to debug in general preferences.

Kind Regards,
NV Access Software Developers

**Comment by lbk2907** [comment]
@SaschaCowley Here I provide the log as you requested. This is the log from the latest NVDA Beta which is version beta-54502,ae1aa96f (2026.1.0.54502). I also include a short audio recording for you to listen just in case you needed.

[nvda.log](https://github.com/user-attachments/files/25198005/nvda.log)
[nvda.mp3](https://github.com/user-attachments/files/25198004/nvda.mp3)

**Comment by valiant8086** [comment]
that is so strange! I definitely can hear the difference. I tried it on my computer and it's exactly the same either way. I wonder if your NVDA.exe has an unusual setting for processor affinity or for the efficiency mode or something similar to that in your task manager that is being overridden when NVDA has a user-facing window open. Example, it's running on an efficiency core when in background or it just is dramatically deprioritized. Also might be the processor scheduling option in the system performance options. I'll sure chime in if I manage to duplicate it somewhere.

**Comment by lbk2907** [comment]
So, may I know it this issue just on my side or anyone also facing this issue as well?
@SaschaCowley @valiant8086 Any suggestion from you?

**Comment by lbk2907** [comment]
I just test on my friend computer and it doesn't happend on his computer. Yeah, strange!
>I wonder if your NVDA.exe has an unusual setting for processor affinity or for the efficiency mode or something similar to that in your task manager that is being overridden when NVDA has a user-facing window open.

@valiant8086 Is there any method to check or test this behavior?

**Comment by lbk2907** [comment]
@SaschaCowley @valiant8086 
Update: I have successfully reproduced the cause of the issue!
1. Priority is the key: In Task Manager > Details, I found that manually setting nvda.exe to "Above Normal" priority immediately fixes the lag, even when no NVDA dialogs are open.
2. Efficiency Mode confirms it: If I turn on Efficiency Mode for nvda.exe, it still have some lag but not as much as I show in the recording above. It still usable and does not go away even if I open an NVDA dialog.
Is there a way for NVDA to automatically request a higher priority or prevent being flagged for "Efficiency Mode" by the OS?

**Comment by valiant8086** [comment]
wow so I actually managed to help, Cool! It shouldn't need to be marked above normal priority like that though. I've never touched that on any computer I've used, and that's a great many of them. I just thought maybe something on your computer, or a user, set it to low priority either experimentally or on purpose, etc. If the community ends up not being able to work out a solution, maybe a tip about adjusting priority belongs in the documentation somewhere to troubleshoot sluggishness.

**Comment by lbk2907** [comment]
@valiant8086 
Yeah, you did it! This help a lot to make things more clearly how to reproduce it.

As for now, the workaround that I use is via this registry key to run the nvda.exe and set the priority as high. Copy the following code into notepad and save the file as .reg instead of .txt and you are good to go. Just run the reg file that you save.
>```
>Windows Registry Editor Version 5.00
>
>[HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows NT\CurrentVersion\Image File Execution Options\nvda.exe\PerfOptions]
>"CpuPriorityClass"=dword:00000003
>
>```

For above normal, just change the value from 3 to 6 in the last line of the code given above.

>```
>"CpuPriorityClass"=dword:00000006
>```

If anyone have better aproach, do share it here. Thanks.

**Comment by lbk2907** [comment]
I am curious, does this issue persist on these newer 'AI-capable' processors like mine?
Quick reference: This is the processor of my laptop.
Intel(R) Core(TM) 5 210H, 2200 Mhz, 8 Core(s), 12 Logical Processor(s)
@SaschaCowley @valiant8086
Do let me know if you need more details or information from my side.

**Comment by josephsl** [comment]
Hi,

I wonder if this is also tied to hardware too? Hybrid processors like the processor described here includes a technology called Intel Thread Director that tells the OS to move threads between performance (P) cores and efficiency (E) cores. For GUI applications, threads with active foreground windows and with active system focus is supposed to receive highest priority possible. This also means Thread Director can tell the OS to schedule high priority threads to performance cores if possible.

While boosting nvda.exe's priority can boost performance, a few things to keep in mind:

1. With performance increase comes increased power usage, and on laptops, shorter battery life.
2. Most of the screen reader's subsystem runs on top of Python runtime. While subsystems like wxPython (that powers NVDA's GUI) is compiled into machine code, the procedures that opens NVDA dialogs, facilitates communication with speech engines, and others are Python bytecode, not strictly machine code. This means boosting nvda.exe's priority also boosts priority of the Python virtual machine that gets called when nvda.exe starts.
3. NVDA, more specifically Python, isn't quite multithreaded (in the strictest sense of the term). While work is under way to make the global interpreter lock (GIL) optional, not all NVDA's subsystems support proper multithreading (some code does support locks and other mechanisms to support multithreading (synchronization across threads, for example)).

Thanks.

**Comment by valiant8086** [comment]
Yup, we're going down the same road. To @lbk2907, try doing without your hack to get the high priority, and set NVDA's process offfinity to something between 0 and 7. That should force it to be stuck only on a P core. If that fixes the problem too, then we might be able to conclude for certain that the real problem is NVDA being forced to run on an E core. According to Gemini, which thinks this is what's going on, that particular CPU uses separate cache for the E cores, and it guesses that the OS may be writing the running cache back and fourth between the E cores one and the P cores one for every interaction. If that's right, then setting the offinity to 11, the last E core, would also make the problem go away although possibly with a very slight reduction in performance. This is in the task manager, so you should be able to do this one too I think, although I'm looking at this on win10 atm and so my taskmgr isn't the same.

I think it should be possible for NVDA to do something about this, but I'm still a little puzzled as to why it's being an issue for that specific environment.

Gemini mentions an application called process lasso which it claims is accessible and has the ability to solve this problem more permanently, if you want to try it. I can't speak for its viability since I can't tell if it does any good, and I wouldn't want to install it to do something I can't tell it's doing lol.

**Comment by lbk2907** [comment]
@josephsl @valiant8086

> Try doing without your hack to get the high priority, and set NVDA's process offfinity to something between 0 and 7.

If I change the priority manually to High, it works perfectly. However, since I have to do this every time NVDA restarts, I use the Registry key to make the High priority permanent.
I also tested the suggestion of changing Process Affinity while keeping the priority at Normal:
- Default (All processors 0-11): The lag is at its worst.
- 0-7: The lag is much better, but there is still a slight, noticeable delay.

In my case, setting the Priority to High is the only way to get 100% snappy performance.

> Gemini mentions an application called process lasso which it claims is accessible and has the ability to solve this problem more permanently, if you want to try it.

I will give this a try and see how it goes.

**Comment by SaschaCowley** [comment]
We plan to investigate how we can work around this in NVDA, for instance by setting NVDA's affinity level, or instructing Windows to schedule NVDA on performance cores.

**Comment by cary-rowen** [comment]
This insights has circulated within the community for many years. The TonysEnhancements add-on is capable of modifying the NVDA process priority, and there are likely other add-ons and registry files being shared as well.
